source/language/openpulse.rst

```diff
@@ -0,0 +1,697 @@
+   extern capture(channel chan, length len, frame output) -> waveform;
+
+  // A capture that returns a count e.g. number of photons detected
+  kernel capture(channel chan, length len, frame output) -> int
+
+The return type of a ``capture`` command varies. It could be a raw trace, ie. a
+list of samples taken over a short period of time. It could be some averaged IQ
+value. It could be a classified bit. Or it could even have no return value,
+pushing the results into some buffer which is then accessed outside the program.
+
+For example, the ``capture`` instruction could return raw waveform data that is then
+discriminated using user-defined boxcar and discrimination ``extern``s.
+
+.. code-block:: javascript
+
+    // Use a boxcar function to generate IQ data from raw waveform
+    extern boxcar(waveform input) -> complex[64];
+    // Use a linear discriminator to generate bits from IQ data
+    extern discriminate(complex[64] iq) -> bit;
+
+    defcal measure $0 -> bit {
+        // Define the channels
+        channel m0 = getchannel("measure", $0);
+        channel cap0 = getchannel("capture", $0);
+
+        // Force time of carrier to 0 for consistent phase for discrimination.
+        frame stimulus_frame = newframe(5e9, 0);
+        frame capture_frame = newframe(5e9, 0);
+
+        // Apply measurement stimulus
+        waveform meas_wf = gaussian_square(1.0, 16000dt, 262dt, 13952dt);
+
+        // Play the stimulus
+        play(m0, meas_wf, stimulus_frame);
+        // Align measure and capture channels
+        barrier(stimulus_frame, capture_frame);
+        // Capture transmitted data after interaction with measurement resonator
+        // extern capture(channel chan, length duration, frame capture_frame) -> waveform;
+        waveform raw_output = capture(cap0, 16000dt, capture_frame);
+
+        // Kernel and discriminate
+        complex[32] iq = boxcar(raw_output);
+        bit result = discriminate(iq);
+
+        return result;
+    }
+
+
+Timing
+------
+
+Each frame maintains its own "clock". When a pulse is played the clock for
+that frame advances by the length of the pulse.
+
+For frames, everything behaves analogous to qubits in the
+`Delays <delays.html>`_ section of this specification. There are however some
+small differences.
+
+The ``delay`` instruction may take a frame instead of a qubit. The ``barrier``
+instruction may also take a list of frames instead of a list of qubits and aligns the time
+of the clocks given as arguments.
+
+``defcal`` blocks have an implicit ``barrier`` on every frame that enters the block,
+meaning that those clocks are guaranteed to be aligned at the start of the block.
+These blocks also need to have a well-defined length, similar to the ``boxas`` block.
+
+.. code-block:: javascript
+
+   cal {
+     waveform p = ...; // some 100dt waveform
+     frame driveframe1 = newframe(5.0e9, 0);
+     frame driveframe2 = newframe(6.0e9, 0);
+   }
+
+   defcal aligned_gates {
```

Comment:
Is this supposed to be just a `cal` block? I don't know what a `defcal` is without qubit arguments.

Reply:
Added qubits in c89f99a.